cmd/snap-repair/trace.go

```diff
@@ -0,0 +1,179 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+// newRepairTraces returns all repairTrace about the given "brand" and "seq"
+// that can be found. brand, seq can be filepath.Glob expressions.
+func newRepairTraces(brand, seq string) ([]*repairTrace, error) {
+	matches, err := filepath.Glob(filepath.Join(dirs.SnapRepairRunDir, brand, seq, "*"))
+	if err != nil {
+		return nil, err
+	}
+
+	var repairTraces []*repairTrace
+	for _, match := range matches {
+		if trace := newRepairTraceFromPath(match); trace != nil {
+			repairTraces = append(repairTraces, trace)
+		}
+	}
+
+	return repairTraces, nil
+}
+
+// repairTrace holds information about a repair that was run.
+type repairTrace struct {
+	path string
+}
+
+// newRepairTraceFromPath takes a repair log path like
+// the path /var/lib/snapd/repair/run/my-brand/1/r2.done
+// and contructs a repair log from that.
+func newRepairTraceFromPath(path string) *repairTrace {
+	rt := &repairTrace{path: path}
+	if !rt.validSuffix(path) {
+		return nil
+	}
+	return rt
+}
+
+// Repair returns the repair human readable string in the form $brand-$id
+func (rt *repairTrace) Repair() string {
+	seq := filepath.Base(filepath.Dir(rt.path))
+	brand := filepath.Base(filepath.Dir(filepath.Dir(rt.path)))
+
+	return fmt.Sprintf("%s-%s", brand, seq)
+}
+
+// Rev returns the revision of the repair
+func (rt *repairTrace) Rev() string {
+	return revFromFilepath(rt.path)
+}
+
+// Summary returns the summary of the repair that was run
+func (rt *repairTrace) Summary() string {
+	f, err := os.Open(rt.path)
+	if err != nil {
+		return "cannot read summary"
+	}
+	defer f.Close()
+
+	needle := "summary: "
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		s := scanner.Text()
```

Comment:
this assumes that summary has no newlines in it, that's reasonable but we need to enforce it then in asserts/repair.go